Stream log output from Kubernetes RunNamespacedJob

joelluijmes opened the pull request:
<!-- Thanks for contributing to Prefect Core! 🎉-->

## Summary
<!-- A sentence summarizing the PR -->

This PR opens a watch on `read_namespaced_pod_log` such that the logging output of the created job is forwarded to prefect. No more black box ⬛!


## Changes
<!-- What does this PR change? -->

Replaces the polling for status on a job, to a watch to forward the logging from the job.


## Importance
<!-- Why is this PR important? -->

Gives insight in what the job is doing.


## Checklist
<!-- PRs will not be reviewed unless these boxes are checked -->

This PR:

- [x] adds new tests (if appropriate)
- [x] adds a change file in the `changes/` directory (if appropriate)
- [x] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

jcrist commented on src/prefect/tasks/kubernetes/job.py:
```diff
@@ -613,9 +614,9 @@ class RunNamespacedJob(Task):
         - kubernetes_api_key_secret (str, optional): the name of the Prefect Secret
             which stored your Kubernetes API Key; this Secret must be a string and in
             BearerToken format
-        - job_status_pool_interval (int, optional): The interval given in seconds
```
Just noticed the typo here (should be `job_status_poll_interval`).

jcrist commented on src/prefect/tasks/kubernetes/job.py:
```diff
@@ -613,9 +614,9 @@ class RunNamespacedJob(Task):
-        - job_status_pool_interval (int, optional): The interval given in seconds
-            indicating how often the Kubernetes API will be requested about the status
-            of the job being performed, defaults to the `5` seconds
+        - log_level (str, optional): Log level used when outputting logs from the job
```
Some jobs will generate lots of logs, and not all users will want logs stored in prefect's backend. This should default to `None` IMO (so it's opt-in).

jcrist commented on src/prefect/tasks/kubernetes/job.py:
```diff
@@ -695,34 +698,68 @@ def run(
+            namespace=namespace, label_selector=pod_selector
+        )
+
+        if len(pods_list.items) != 1:
```
Some jobs might run multiple pods

joelluijmes replied:
Hmm, I was thinking about this, but I thought it was impossible, how can a job create multiple pods?

jcrist replied:
By specifying `parallelism` or `completions`: https://kubernetes.io/docs/concepts/workloads/controllers/job/#parallel-jobs

jcrist commented on src/prefect/tasks/kubernetes/job.py:
```diff
@@ -695,34 +698,68 @@ def run(
+
+        self.logger.info(f"Pod {pod_name} is running.")
+
+        # The watch stream is automatically stopped when the job is completed or failed. So we don't
+        # have to poll the job its status.
```
Watch streams can complete for other reasons (stale event requiring a restart, api error, etc...). Most code I've used them required an extra loop around `stream` calls to handle restarts/spurious errors.

jcrist commented on src/prefect/tasks/kubernetes/job.py:
```diff
@@ -695,34 +698,68 @@ def run(
         self.logger.info(f"Job {job_name} has been created.")
 
-        completed = False
-        while not completed:
-            res = api_client.read_namespaced_job_status(
-                name=job_name, namespace=namespace
```
Since:
- not all users will want to stream logs
- relying on a log stream for job completion isn't sufficient

I think we'll still want to check for job completion via the job status. The polling loop is one way, starting a watch stream on the job status would be another (I could go either way here).

jcrist reviewed (changes requested):
Thanks @joelluijmes. I agree that streaming the logs from a job into prefect sounds useful, but I think we'll want to make it opt-in.

joelluijmes commented:
> Thanks @joelluijmes. I agree that streaming the logs from a job into prefect sounds useful, but I think we'll want to make it opt-in.

Thanks for the quick review, I'll update the PR tomorrow with requested chagnes 👍

joelluijmes commented:
Just worked on it, and reimplemented the polling loop in order to support multiple pods. However I'm now running into how `read_namespaced_pod_log` works, ideally we want to have pagination. Or an interruptible stream (such that it doesn't keep hanging). 

I tried to paginate the logs with `since_seconds` but that is not precise enough and skips logs if multiple logs are  happen in milliseconds. I could add an overlap second, but then we have to filter on duplicate lines but that is tricky because we don't know the job is making duplicates or not.

The possible solution as I now see:
1. Don't stream the logs, only retrieve them on completion (most easy, but I'd like to see the job logs in action). 
2. For each pod start a watch in a new thread. 

Do you have any suggestions or insights on this? 

--

I'll try to see what I can do with 2, honestly I always shudder a bit with threads as I'm used to co-routines. But I see that api_client of kubernetes internally also use a ThreadPool when param `async_req=True`, so maybe that is a direction I could take a look at.

joshmeek commented:
Hey @joelluijmes would you mind merging the latest from master into your branch? We merged in a PR yesterday that fixes some of the issues we have been seeing in the CI tests!

joelluijmes commented:
> Hey @joelluijmes would you mind merging the latest from master into your branch? We merged in a PR yesterday that fixes some of the issues we have been seeing in the CI tests!

Sure! I reverted the merge because I suspected it as the cause for the build failing 😬

joelluijmes commented:
Thought the diff was weird after the master merge. Probably did something wrong 🤷 
Anyhow, rebased the branch on to latest master and squashed into meaningful commits 😄 

Should be ready for review.

joelluijmes commented:
Any remarks on this?

jcrist commented on src/prefect/tasks/kubernetes/job.py:
```diff
@@ -695,34 +709,76 @@ def run(
+                        )
+
+                        self.logger.info(f"Started following logs for {pod_name}")
+                        pod_log_streams[pod_name] = pool.apply_async(read_pod_logs.run)
```
Eventually the thread pool will fill up, so active streaming of logs will only work for jobs with n-pods <= n-threads. This is a bit unfortunate - it'd be easy to support more if we were using an async kubernetes client but we're not.

I think I'm fine with this for now, but in the future I may want to fix this.

jcrist replied:
Actually, would you mind adding a todo comment above this line with that info? Something like

```
# TODO: if a job has more active pods than the threadpool has threads
# logs will no longer be streamed live for all pods. Using an async
# k8s client would help remedy this.
```

joelluijmes replied:
Aha alright, I'm not that well versed in Python multiprocessing, but good remark, I'll add it! An async kubernetes client would probably make this bit easier 😬 

To what size does Python's threadpool grow?

jcrist replied:
> To what size does Python's threadpool grow?

It's static, and equal to the number of cores.

Actually we might want to switch to a [`concurrent.futures.TheadPoolExecutor`](https://docs.python.org/3/library/concurrent.futures.html#threadpoolexecutor) instead. This will grow in size but starts no threads by default (the default cap is `min(32, (os.cpu_count() or 1) + 4)`). You'd use `submit` instead of `apply_async` and `shutdown` instead of `close`, but everything else should be equivalent.

jcrist commented on src/prefect/tasks/kubernetes/pod.py:
```diff
@@ -571,3 +573,118 @@ def run(
+        - pod_name (str, optional): The name of a pod to replace
+        - namespace (str, optional): The Kubernetes namespace to patch this pod in,
+            defaults to the `default` namespace
+        - func_stream (Callable, optional): Stream the logs to provided function,
+            for each line the function is called.
```
I'd find the following kwarg name (and docstring) clearer. Same in the `run` method below.

```
        - on_log_entry (Callable, optional): If provided, will stream the pod logs
            calling the callback for every line (and the task returns `None`). If not
            provided, the current pod logs will be returned immediately from the task.
```

jcrist commented on src/prefect/tasks/kubernetes/pod.py:
```diff
@@ -571,3 +573,118 @@ def run(
         )
+
+
+class ReadNamespacedPodLogs(Task):
```
Can you add this to the api docs (add to the list of classes [here](https://github.com/PrefectHQ/prefect/blob/928d2b9e9b54847e3d021a544c9cdeb8ea1ba10b/docs/outline.toml#L312)).

jcrist reviewed (commented):
Apologies for the delayed review, this is looking pretty good! Only a few small fixes needed.

joelluijmes commented:
> Apologies for the delayed review, this is looking pretty good! Only a few small fixes needed.

No worries, just like to see this merged so I can use it 🤓 
Great, updated with the requested changes!

joelluijmes commented:
Okay, better late then never, but completed the requested changes 👍 

One thing we may need to look out for is following. The kubernetes library list this as comment on the stream
```
Note that watching an API resource can expire. The method tries to
        resume automatically once from the last result, but if that last result
        is too old as well, an `ApiException` exception will be thrown with
        ``code`` 410. In that case you have to recover yourself, probably
        by listing the API resource to obtain the latest state and then
        watching from that state on by setting ``resource_version`` to
        one returned from listing.
```

However, when passing `resource_version`, None or the actual value, the generator doesn't yield any logs. Maybe something is broken in the Kubernetes library, or this is just not supported for `read_namespaced_pod_log`. In my experience, it just doesn't work. I tried it with two different Kubernetes clusters (GKE at 1.19 and a bare metal at 1.17).

joelluijmes commented:
Just noticed that the build wasn't successful after my last commit. I've rebased it on latest master and now it's green. Hope it can be merged soon 👍

jcrist commented:
Oop, apologies for letting this linger. I'll try to finish this up today and get it in. Thanks @joelluijmes!
